pkg/progress/progress.go

```diff
@@ -62,52 +78,89 @@ func (m *Manager) Reset() {
 	m.progresses = make(map[string]*progressIndicator)
 }
 
+// Option is used to do some action for progressIndicator.
+type Option func(*progressIndicator)
+
+// WindowDurationOption changes the time window size.
+func WindowDurationOption(dur time.Duration) func(*progressIndicator) {
+	return func(pi *progressIndicator) {
+		if dur < minSpeedCalculationWindow {
+			dur = minSpeedCalculationWindow
+		} else if dur > maxSpeedCalculationWindow {
+			dur = maxSpeedCalculationWindow
+		}
+		pi.windowLength = int(dur/pi.updateInterval) + 1
+	}
+}
+
 // AddProgress adds a progress into manager if it doesn't exist.
-func (m *Manager) AddProgress(progress string, current, total float64, updateInterval time.Duration) (exist bool) {
+func (m *Manager) AddProgress(progress string, current, total float64, updateInterval time.Duration, opts ...Option) (exist bool) {
 	m.Lock()
 	defer m.Unlock()
 
 	history := list.New()
 	history.PushBack(current)
 	if _, exist = m.progresses[progress]; !exist {
-		m.progresses[progress] = &progressIndicator{
-			total:             total,
-			remaining:         total,
-			history:           history,
-			windowLengthLimit: int(speedStatisticalWindow / updateInterval),
-			updateInterval:    updateInterval,
+		pi := &progressIndicator{
+			total:          total,
+			remaining:      total,
+			history:        history,
+			windowCapacity: int(maxSpeedCalculationWindow/updateInterval) + 1,
+			windowLength:   int(minSpeedCalculationWindow / updateInterval),
+			updateInterval: updateInterval,
+		}
+		for _, op := range opts {
+			op(pi)
 		}
+		m.progresses[progress] = pi
+		pi.front = history.Front()
+		pi.position = 1
 	}
 	return
 }
 
 // UpdateProgress updates the progress if it exists.
-func (m *Manager) UpdateProgress(progress string, current, remaining float64, isInc bool) {
+func (m *Manager) UpdateProgress(progress string, current, remaining float64, isInc bool, opts ...Option) {
 	m.Lock()
 	defer m.Unlock()
 
 	if p, exist := m.progresses[progress]; exist {
+		for _, op := range opts {
+			op(p)
+		}
 		p.remaining = remaining
 		if p.total < remaining {
 			p.total = remaining
 		}
 
-		if p.history.Len() > p.windowLengthLimit {
+		p.history.PushBack(current)
+		p.position++
+
+		// try to move `front` into correct place.
+		for p.position > p.windowLength {
+			p.front = p.front.Next()
+			p.position--
+		}
+		for p.position < p.windowLength && p.front.Prev() != nil {
```

Comment:
Will it only happen in the beginning? The `position` here is confusing to me.

Reply:
No. If `windowLength` becomes large, the `front` should move forward

Reply:
added some comments.